File: shuffle.php
<?php
// shuffle.php - Перемешиватель списков на PHP (CLI + веб)
// CLI: php shuffle.php --list "1,2,3,4,5" --algorithm 1
// Веб: откройте как HTML

// ========== АЛГОРИТМЫ ПЕРЕМЕШИВАНИЯ ==========
function fisherYates($arr, $rng) {
    $result = $arr;
    for ($i = count($result) - 1; $i > 0; $i--) {
        $j = $rng(0, $i);
        $tmp = $result[$i];
        $result[$i] = $result[$j];
        $result[$j] = $tmp;
    }
    return $result;
}

function sortKeyShuffle($arr, $rng) {
    $result = $arr;
    usort($result, function($a, $b) use ($rng) {
        return $rng(0, 1) < 0.5 ? -1 : 1;
    });
    return $result;
}

function builtinShuffle($arr, $rng) {
    return fisherYates($arr, $rng);
}

$algorithms = [
    '1' => ['name' => 'Fisher-Yates (оптимальный)', 'func' => 'fisherYates'],
    '2' => ['name' => 'Сортировка случайным ключом', 'func' => 'sortKeyShuffle'],
    '3' => ['name' => 'Встроенный (стандартный)', 'func' => 'builtinShuffle'],
];

// ========== СТАТИСТИКА ==========
function countInversions($arr) {
    $inv = 0;
    $n = count($arr);
    for ($i = 0; $i < $n; $i++) {
        for ($j = $i + 1; $j < $n; $j++) {
            if ($arr[$i] > $arr[$j]) $inv++;
        }
    }
    return $inv;
}

function entropy($arr) {
    if (empty($arr) || !is_numeric($arr[0])) return 0.0;
    $min = min($arr);
    $max = max($arr);
    if ($max == $min) return 0.0;
    $bins = 10;
    $hist = array_fill(0, $bins, 0);
    foreach ($arr as $x) {
        $idx = (int)(($x - $min) / ($max - $min) * $bins);
        if ($idx == $bins) $idx = $bins - 1;
        $hist[$idx]++;
    }
    $ent = 0.0;
    $n = count($arr);
    foreach ($hist as $count) {
        if ($count > 0) {
            $p = $count / $n;
            $ent -= $p * log($p, 2);
        }
    }
    return $ent;
}

function getStats($arr) {
    return [
        'length' => count($arr),
        'inversions' => countInversions($arr),
        'entropy' => entropy($arr),
    ];
}

// ========== ВВОД/ВЫВОД ==========
function parseList($input) {
    $parts = array_map('trim', explode(',', $input));
    $result = [];
    foreach ($parts as $p) {
        if (is_numeric($p)) {
            if (strpos($p, '.') !== false) $result[] = (float)$p;
            else $result[] = (int)$p;
        } else {
            $result[] = $p;
        }
    }
    return $result;
}

function formatList($arr) {
    return '[' . implode(', ', array_map(function($x) { return is_string($x) ? '"' . $x . '"' : $x; }, $arr)) . ']';
}

function exportJSON($arr, $filename) {
    file_put_contents($filename, json_encode($arr, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
}

function exportCSV($arr, $filename) {
    $f = fopen($filename, 'w');
    fputcsv($f, ['value']);
    foreach ($arr as $val) fputcsv($f, [$val]);
    fclose($f);
}

function importList($filename) {
    $content = file_get_contents($filename);
    if (strpos($filename, '.json') !== false) {
        return json_decode($content, true);
    } else {
        $lines = array_map('trim', explode("\n", $content));
        $result = [];
        for ($i = 1; $i < count($lines); $i++) {
            if ($lines[$i] !== '') $result[] = $lines[$i];
        }
        return $result;
    }
}

function generateRandomList($size, $minVal = 0, $maxVal = 100) {
    $result = [];
    for ($i = 0; $i < $size; $i++) {
        $result[] = rand($minVal, $maxVal);
    }
    return $result;
}

function getInput($prompt) {
    echo $prompt;
    return trim(fgets(STDIN));
}

if (php_sapi_name() === 'cli') {
    $options = getopt("", ["list:", "file:", "random:", "algorithm:", "seed:", "export:", "stats"]);
    if (isset($options['list']) || isset($options['file']) || isset($options['random'])) {
        $data = [];
        if (isset($options['list'])) {
            $data = parseList($options['list']);
        } elseif (isset($options['file'])) {
            $data = importList($options['file']);
        } elseif (isset($options['random'])) {
            $size = (int)$options['random'];
            $data = generateRandomList($size);
        }
        if (empty($data)) { echo "Список пуст.\n"; exit; }
        $algoKey = isset($options['algorithm']) ? $options['algorithm'] : '1';
        if (!isset($algorithms[$algoKey])) $algoKey = '1';
        $func = $algorithms[$algoKey]['func'];
        $rng = function($min, $max) {
            static $seed = null;
            if ($seed === null && isset($options['seed'])) {
                $seed = (int)$options['seed'];
                srand($seed);
            }
            return rand($min, $max);
        };
        if (isset($options['seed'])) {
            $seed = (int)$options['seed'];
            srand($seed);
            $rng = function($min, $max) use ($seed) {
                return rand($min, $max);
            };
        } else {
            $rng = function($min, $max) {
                return rand($min, $max);
            };
        }
        $shuffled = $func($data, $rng);
        echo "Исходный список: " . formatList($data) . "\n";
        echo "Перемешанный список: " . formatList($shuffled) . "\n";
        if (isset($options['stats'])) {
            $stats = getStats($shuffled);
            echo "\n📊 Статистика:\n";
            echo "  Длина: {$stats['length']}\n";
            echo "  Инверсий: {$stats['inversions']}\n";
            echo "  Энтропия: " . number_format($stats['entropy'], 4) . "\n";
        }
        if (isset($options['export'])) {
            if (strpos($options['export'], '.json') !== false) exportJSON($shuffled, $options['export']);
            else exportCSV($shuffled, $options['export']);
            echo "Экспортировано в {$options['export']}\n";
        }
    } else {
        // Интерактивный режим
        echo "🎲 ПЕРЕМЕШИВАТЕЛЬ СПИСКОВ\n";
        echo "Введите список (через запятую):\n";
        $input = getInput("> ");
        if (empty($input)) {
            echo "Список пуст. Сгенерируем случайный.\n";
            $size = (int)getInput("Размер случайного списка (по умолчанию 10): ") ?: 10;
            $data = generateRandomList($size);
            echo "Сгенерированный список: " . formatList($data) . "\n";
        } else {
            $data = parseList($input);
            echo "Исходный список: " . formatList($data) . "\n";
        }

        echo "\nВыберите алгоритм:\n";
        foreach ($algorithms as $key => $alg) {
            echo "$key. {$alg['name']}\n";
        }
        $algoChoice = getInput("Ваш выбор (по умолчанию 1): ") ?: '1';
        if (!isset($algorithms[$algoChoice])) $algoChoice = '1';
        $func = $algorithms[$algoChoice]['func'];

        $seedInput = getInput("Введите seed (или оставьте пустым для случайного): ");
        if ($seedInput !== '') {
            $seed = (int)$seedInput;
            srand($seed);
            $rng = function($min, $max) { return rand($min, $max); };
        } else {
            $rng = function($min, $max) { return rand($min, $max); };
        }

        $shuffled = $func($data, $rng);
        echo "\nПеремешанный список: " . formatList($shuffled) . "\n";

        $stats = getStats($shuffled);
        echo "\n📊 Статистика:\n";
        echo "  Длина: {$stats['length']}\n";
        echo "  Инверсий: {$stats['inversions']}\n";
        echo "  Энтропия: " . number_format($stats['entropy'], 4) . "\n";

        $export = getInput("\nЭкспортировать результат? (y/n): ");
        if (strtolower($export) == 'y') {
            $fmt = getInput("Формат (json/csv): ");
            $filename = getInput("Имя файла: ");
            if (empty($filename)) $filename = "shuffled.$fmt";
            if ($fmt == 'json') exportJSON($shuffled, $filename);
            else exportCSV($shuffled, $filename);
            echo "Экспортировано в $filename\n";
        }
    }
    exit;
}

// ========== ВЕБ-ИНТЕРФЕЙС ==========
?>
<!DOCTYPE html>
<html lang="ru">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>🎲 Перемешиватель списков (PHP)</title>
    <style>
        body { font-family: 'Segoe UI', sans-serif; background: #f4f7fb; margin: 20px; }
        .container { max-width: 800px; margin: 0 auto; background: white; padding: 20px; border-radius: 16px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }
        h1 { text-align: center; }
        .form-group { margin-bottom: 15px; }
        label { display: inline-block; width: 120px; }
        input, select, button { padding: 6px; border-radius: 4px; border: 1px solid #ccc; }
        button { background: #3498db; color: white; border: none; cursor: pointer; padding: 6px 20px; }
        button:hover { background: #2980b9; }
        .result { background: #ecf0f1; padding: 15px; border-radius: 8px; margin-top: 20px; }
        .stats { font-family: monospace; }
    </style>
</head>
<body>
<div class="container">
    <h1>🎲 Перемешиватель списков (PHP)</h1>
    <form method="GET">
        <div class="form-group">
            <label>Список чисел:</label>
            <input type="text" name="list" placeholder="1,2,3,4,5" value="<?= isset($_GET['list']) ? htmlspecialchars($_GET['list']) : '' ?>">
        </div>
        <div class="form-group">
            <label>Алгоритм:</label>
            <select name="algorithm">
                <?php foreach ($algorithms as $key => $alg): ?>
                    <option value="<?= $key ?>" <?= isset($_GET['algorithm']) && $_GET['algorithm'] == $key ? 'selected' : '' ?>><?= $alg['name'] ?></option>
                <?php endforeach; ?>
            </select>
        </div>
        <div class="form-group">
            <label>Seed (опционально):</label>
            <input type="number" name="seed" value="<?= isset($_GET['seed']) ? htmlspecialchars($_GET['seed']) : '' ?>">
        </div>
        <div class="form-group">
            <label>Показать статистику:</label>
            <input type="checkbox" name="stats" <?= isset($_GET['stats']) ? 'checked' : '' ?>>
        </div>
        <button type="submit">Перемешать</button>
        <a href="?export_csv=1&<?= http_build_query($_GET) ?>">📥 CSV</a>
        <a href="?export_json=1&<?= http_build_query($_GET) ?>">📥 JSON</a>
    </form>

    <?php if (isset($_GET['export_csv']) || isset($_GET['export_json'])): 
        $list = $_GET['list'] ?? '';
        if ($list) {
            $data = parseList($list);
            if ($data) {
                $algoKey = $_GET['algorithm'] ?? '1';
                if (!isset($algorithms[$algoKey])) $algoKey = '1';
                $func = $algorithms[$algoKey]['func'];
                $seed = isset($_GET['seed']) ? (int)$_GET['seed'] : null;
                if ($seed !== null) srand($seed);
                $rng = function($min, $max) { return rand($min, $max); };
                $shuffled = $func($data, $rng);
                if (isset($_GET['export_csv'])) {
                    header('Content-Type: text/csv');
                    header('Content-Disposition: attachment; filename="shuffled.csv"');
                    exportCSV($shuffled, 'php://output');
                    exit;
                } else {
                    header('Content-Type: application/json');
                    echo json_encode($shuffled, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);
                    exit;
                }
            }
        }
    endif; ?>

    <?php if (isset($_GET['list']) && !empty($_GET['list'])): 
        $data = parseList($_GET['list']);
        if ($data):
            $algoKey = $_GET['algorithm'] ?? '1';
            if (!isset($algorithms[$algoKey])) $algoKey = '1';
            $func = $algorithms[$algoKey]['func'];
            $seed = isset($_GET['seed']) ? (int)$_GET['seed'] : null;
            if ($seed !== null) srand($seed);
            $rng = function($min, $max) { return rand($min, $max); };
            $shuffled = $func($data, $rng);
    ?>
        <div class="result">
            <p><strong>Исходный список:</strong> <?= formatList($data) ?></p>
            <p><strong>Перемешанный список:</strong> <?= formatList($shuffled) ?></p>
            <?php if (isset($_GET['stats'])): 
                $stats = getStats($shuffled);
            ?>
                <div class="stats">
                    <h4>📊 Статистика:</h4>
                    <p>Длина: <?= $stats['length'] ?></p>
                    <p>Инверсий: <?= $stats['inversions'] ?></p>
                    <p>Энтропия: <?= number_format($stats['entropy'], 4) ?></p>
                </div>
            <?php endif; ?>
        </div>
    <?php endif; endif; ?>
</div>
</body>
</html>

File: shuffle.go
// shuffle.go - Перемешиватель списков на Go (CLI)
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

// ========== АЛГОРИТМЫ ПЕРЕМЕШИВАНИЯ ==========
func fisherYatesShuffle(arr []interface{}, rng *rand.Rand) []interface{} {
	result := make([]interface{}, len(arr))
	copy(result, arr)
	for i := len(result) - 1; i > 0; i-- {
		j := rng.Intn(i + 1)
		result[i], result[j] = result[j], result[i]
	}
	return result
}

func sortKeyShuffle(arr []interface{}, rng *rand.Rand) []interface{} {
	result := make([]interface{}, len(arr))
	copy(result, arr)
	// Сортируем с использованием случайного ключа
	sort.Slice(result, func(i, j int) bool {
		return rng.Float64() < 0.5
	})
	return result
}

func builtinShuffle(arr []interface{}, rng *rand.Rand) []interface{} {
	// Используем Fisher-Yates
	return fisherYatesShuffle(arr, rng)
}

type Algorithm struct {
	Name string
	Func func([]interface{}, *rand.Rand) []interface{}
}

var algorithms = map[string]Algorithm{
	"1": {"Fisher-Yates (оптимальный)", fisherYatesShuffle},
	"2": {"Сортировка случайным ключом", sortKeyShuffle},
	"3": {"Встроенный (стандартный)", builtinShuffle},
}

// ========== СТАТИСТИКА ==========
func countInversions(arr []interface{}) int {
	inv := 0
	for i := 0; i < len(arr); i++ {
		for j := i + 1; j < len(arr); j++ {
			// Сравнение только для чисел
			if vi, ok := arr[i].(float64); ok {
				if vj, ok := arr[j].(float64); ok {
					if vi > vj {
						inv++
					}
				}
			}
		}
	}
	return inv
}

func entropy(arr []interface{}) float64 {
	if len(arr) == 0 {
		return 0.0
	}
	// Проверяем, все ли числа
	numbers := make([]float64, 0, len(arr))
	for _, v := range arr {
		if f, ok := v.(float64); ok {
			numbers = append(numbers, f)
		} else {
			return 0.0
		}
	}
	if len(numbers) == 0 {
		return 0.0
	}
	minVal := numbers[0]
	maxVal := numbers[0]
	for _, v := range numbers {
		if v < minVal {
			minVal = v
		}
		if v > maxVal {
			maxVal = v
		}
	}
	if maxVal == minVal {
		return 0.0
	}
	bins := 10
	hist := make([]int, bins)
	for _, v := range numbers {
		idx := int((v - minVal) / (maxVal - minVal) * float64(bins))
		if idx == bins {
			idx = bins - 1
		}
		hist[idx]++
	}
	ent := 0.0
	n := float64(len(numbers))
	for _, count := range hist {
		if count > 0 {
			p := float64(count) / n
			ent -= p * math.Log2(p)
		}
	}
	return ent
}

func getStats(arr []interface{}) map[string]interface{} {
	return map[string]interface{}{
		"length":     len(arr),
		"inversions": countInversions(arr),
		"entropy":    entropy(arr),
	}
}

// ========== ВВОД/ВЫВОД ==========
func parseList(input string) []interface{} {
	parts := strings.Split(input, ",")
	result := make([]interface{}, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		if val, err := strconv.Atoi(p); err == nil {
			result = append(result, float64(val))
		} else if val, err := strconv.ParseFloat(p, 64); err == nil {
			result = append(result, val)
		} else {
			result = append(result, p)
		}
	}
	return result
}

func formatList(arr []interface{}) string {
	parts := make([]string, len(arr))
	for i, v := range arr {
		switch v := v.(type) {
		case float64:
			if v == float64(int(v)) {
				parts[i] = fmt.Sprintf("%d", int(v))
			} else {
				parts[i] = fmt.Sprintf("%.2f", v)
			}
		default:
			parts[i] = fmt.Sprintf("%v", v)
		}
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

func exportJSON(arr []interface{}, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	return encoder.Encode(arr)
}

func exportCSV(arr []interface{}, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{"value"})
	for _, v := range arr {
		writer.Write([]string{fmt.Sprintf("%v", v)})
	}
	return nil
}

func importList(filename string) ([]interface{}, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	if strings.HasSuffix(filename, ".json") {
		var arr []interface{}
		if err := json.Unmarshal(data, &arr); err != nil {
			return nil, err
		}
		return arr, nil
	} else {
		// CSV
		lines := strings.Split(string(data), "\n")
		var arr []interface{}
		for i, line := range lines {
			if i == 0 {
				continue // пропускаем заголовок
			}
			line = strings.TrimSpace(line)
			if line == "" {
				continue
			}
			if val, err := strconv.ParseFloat(line, 64); err == nil {
				arr = append(arr, val)
			} else {
				arr = append(arr, line)
			}
		}
		return arr, nil
	}
}

func generateRandomList(size int, minVal, maxVal int) []interface{} {
	arr := make([]interface{}, size)
	for i := 0; i < size; i++ {
		arr[i] = float64(rand.Intn(maxVal-minVal+1) + minVal)
	}
	return arr
}

// ========== ОСНОВНАЯ ЛОГИКА ==========
func interactive() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("🎲 ПЕРЕМЕШИВАТЕЛЬ СПИСКОВ")
	fmt.Println("Введите список (через запятую):")
	fmt.Print("> ")
	scanner.Scan()
	input := scanner.Text()
	var data []interface{}
	if input == "" {
		fmt.Print("Размер случайного списка (по умолчанию 10): ")
		scanner.Scan()
		sizeStr := scanner.Text()
		size := 10
		if sizeStr != "" {
			size, _ = strconv.Atoi(sizeStr)
		}
		data = generateRandomList(size, 0, 100)
		fmt.Printf("Сгенерированный список: %s\n", formatList(data))
	} else {
		data = parseList(input)
		fmt.Printf("Исходный список: %s\n", formatList(data))
	}

	fmt.Println("\nВыберите алгоритм:")
	for key, alg := range algorithms {
		fmt.Printf("%s. %s\n", key, alg.Name)
	}
	fmt.Print("Ваш выбор (по умолчанию 1): ")
	scanner.Scan()
	algoChoice := scanner.Text()
	if algoChoice == "" || algorithms[algoChoice].Name == "" {
		algoChoice = "1"
	}
	algorithm := algorithms[algoChoice].Func

	fmt.Print("Введите seed (или оставьте пустым для случайного): ")
	scanner.Scan()
	seedStr := scanner.Text()
	var rng *rand.Rand
	if seedStr != "" {
		seed, err := strconv.ParseInt(seedStr, 10, 64)
		if err != nil {
			fmt.Println("Неверный seed, используется случайный.")
			rng = rand.New(rand.NewSource(time.Now().UnixNano()))
		} else {
			rng = rand.New(rand.NewSource(seed))
		}
	} else {
		rng = rand.New(rand.NewSource(time.Now().UnixNano()))
	}

	shuffled := algorithm(data, rng)
	fmt.Printf("\nПеремешанный список: %s\n", formatList(shuffled))

	stats := getStats(shuffled)
	fmt.Println("\n📊 Статистика:")
	fmt.Printf("  Длина: %d\n", stats["length"])
	fmt.Printf("  Инверсий: %d\n", stats["inversions"])
	fmt.Printf("  Энтропия: %.4f\n", stats["entropy"])

	fmt.Print("\nЭкспортировать результат? (y/n): ")
	scanner.Scan()
	exportChoice := scanner.Text()
	if exportChoice == "y" || exportChoice == "Y" {
		fmt.Print("Формат (json/csv): ")
		scanner.Scan()
		fmt.Print("Имя файла: ")
		scanner.Scan()
		filename := scanner.Text()
		if filename == "" {
			filename = "shuffled." + scanner.Text()
		}
		var err error
		if scanner.Text() == "json" {
			err = exportJSON(shuffled, filename)
		} else {
			err = exportCSV(shuffled, filename)
		}
		if err != nil {
			fmt.Printf("Ошибка: %v\n", err)
		} else {
			fmt.Printf("Экспортировано в %s\n", filename)
		}
	}
}

func cli() {
	var list string
	var file string
	var random int
	var algorithmKey string
	var seed int64
	var export string
	var statsFlag bool

	flag.StringVar(&list, "list", "", "Список через запятую")
	flag.StringVar(&file, "file", "", "Файл для импорта (JSON/CSV)")
	flag.IntVar(&random, "random", 0, "Сгенерировать случайный список")
	flag.StringVar(&algorithmKey, "algorithm", "1", "Алгоритм (1-3)")
	flag.Int64Var(&seed, "seed", 0, "Seed для воспроизводимости")
	flag.StringVar(&export, "export", "", "Экспорт в файл")
	flag.BoolVar(&statsFlag, "stats", false, "Показать статистику")
	flag.Parse()

	var data []interface{}
	if list != "" {
		data = parseList(list)
	} else if file != "" {
		var err error
		data, err = importList(file)
		if err != nil {
			fmt.Printf("Ошибка импорта: %v\n", err)
			return
		}
	} else if random > 0 {
		data = generateRandomList(random, 0, 100)
	} else {
		fmt.Println("Укажите --list, --file или --random")
		return
	}
	if len(data) == 0 {
		fmt.Println("Список пуст.")
		return
	}

	var rng *rand.Rand
	if seed != 0 {
		rng = rand.New(rand.NewSource(seed))
	} else {
		rng = rand.New(rand.NewSource(time.Now().UnixNano()))
	}

	algo, ok := algorithms[algorithmKey]
	if !ok {
		algo = algorithms["1"]
	}
	shuffled := algo.Func(data, rng)

	fmt.Printf("Исходный список: %s\n", formatList(data))
	fmt.Printf("Перемешанный список: %s\n", formatList(shuffled))
	if statsFlag {
		stats := getStats(shuffled)
		fmt.Println("\n📊 Статистика:")
		fmt.Printf("  Длина: %d\n", stats["length"])
		fmt.Printf("  Инверсий: %d\n", stats["inversions"])
		fmt.Printf("  Энтропия: %.4f\n", stats["entropy"])
	}
	if export != "" {
		var err error
		if strings.HasSuffix(export, ".json") {
			err = exportJSON(shuffled, export)
		} else {
			err = exportCSV(shuffled, export)
		}
		if err != nil {
			fmt.Printf("Ошибка экспорта: %v\n", err)
		} else {
			fmt.Printf("Экспортировано в %s\n", export)
		}
	}
}

func main() {
	if len(os.Args) > 1 {
		cli()
	} else {
		interactive()
	}
}
